substratevm/src/com.oracle.svm.agent/src/com/oracle/svm/agent/BreakpointInterceptor.java

```diff
@@ -678,6 +699,124 @@ private static boolean handleGetSystemResources(JNIEnvironment jni, Breakpoint b
+            postDefineCLassAccessAdvisor.setInLivePhase(true);
+            if (postDefineCLassAccessAdvisor.shouldIgnore(new LazyValue<>(() -> definedClassName), new LazyValue<>(() -> null))) {
+                isDynamicallyGenerated = false;
+            }
+
+            // 2. Class with name starts with $ or contains $$ is usually dynamically generated
+            String className = definedClassName.substring(definedClassName.lastIndexOf('.') + 1);
+            if (className.startsWith("$") || className.contains("$$")) {
+                isDynamicallyGenerated = true;
+            } else {
+                // 3. A dynamically defined class always return null
+                // when call java.lang.ClassLoader.getResource(classname)
+                // This is the accurate but slow way.
+                String asResourceName = definedClassName.replace('.', '/') + ".class";
+                try (CCharPointerHolder resourceNameHolder = toCString(asResourceName);) {
+                    JNIObjectHandle resourceNameJString = jniFunctions().getNewStringUTF().invoke(jni, resourceNameHolder.get());
+                    JNIObjectHandle returnValue = callObjectMethodL(jni, self, agent.handles().javaLangClassLoaderGetResource, resourceNameJString);
+                    isDynamicallyGenerated = returnValue.equal(nullHandle());
+                }
+            }
+        }
+
+        // Reuse verifyForName
+        // CallerClass is always java.lang.ClassLoader, we only check the defined class
+        boolean allowed = (accessVerifier == null || accessVerifier.verifyForName(jni, nullHandle(), definedClassName));
+        Object result = false;
+        boolean justAdded = definedClasses.add(definedClassName);
+        if (isDynamicallyGenerated) {
+            if (!justAdded) {
+                unsupportedExceptions.add("Class " + definedClassName + " has been defined before. Multiple definitions are not supported.\n" +
+                                AbstractDynamicClassGenerationSupport.getStackTrace(jni).toString());
+                return allowed;
+            }
+            // Check the caller's protectionDomain is null, otherwise don't support
+            String caller = getMethodFullNameAtFrame(jni, 1);
+            if ("java.lang.ClassLoader.defineClass(Ljava/lang/String;[BIILjava/security/ProtectionDomain;)Ljava/lang/Class;".equals(caller)) {
+                caller = getMethodFullNameAtFrame(jni, 2);
+                if (!"java.lang.ClassLoader.defineClass(Ljava/lang/String;[BII)Ljava/lang/Class;".equals(caller)) {
+                    // Reflections are supported in another way, no need to track dynamic class
+                    // loading
+                    // used by reflection.
+                    if (!caller.startsWith("sun.reflect")) {
+                        unsupportedExceptions.add("Don't support defineClass with ProtectionDomain parameter in native image. " +
```

Comment:
I don't see why we would fail here. I think we should simply not collect those classes ATM. Actually, we could even collect them because in the future we definitely want to support the calls with the protection domain.

Reply:
Reporting at collecting time would let the user know the program is not supported as early as possible, so they don't have to wait to image compilation time or runtime. This unsupported message is reported at unLoad time so it wouldn't interfere the configuration collecting process.

Reply:
One of the use cases for the agent is automatic collection of profiles in the cloud setting. For this purpose, I would like to have the agent behave exactly the same as the original. 

Could we put this kind of reporting behind a flag?

Reply:
removed the reporting here.